adagrams/game.py

```diff
@@ -1,11 +1,142 @@
+from random import sample
+from collections import Counter
+# import pprint
+
+
+LETTER_POOL = {
+    'A': 9, 
+    'B': 2, 
+    'C': 2, 
+    'D': 4, 
+    'E': 12, 
+    'F': 2, 
+    'G': 3, 
+    'H': 2, 
+    'I': 9, 
+    'J': 1, 
+    'K': 1, 
+    'L': 4, 
+    'M': 2, 
+    'N': 6, 
+    'O': 8, 
+    'P': 2, 
+    'Q': 1, 
+    'R': 6, 
+    'S': 4, 
+    'T': 6, 
+    'U': 4, 
+    'V': 2, 
+    'W': 2, 
+    'X': 1, 
+    'Y': 2, 
+    'Z': 1
+}
+
+SCORE_CHART = {
+    'A': 1, 
+    'B': 3, 
+    'C': 3, 
+    'D': 2, 
+    'E': 1, 
+    'F': 4, 
+    'G': 2, 
+    'H': 4, 
+    'I': 1, 
+    'J': 8, 
+    'K': 5, 
+    'L': 1, 
+    'M': 3, 
+    'N': 1, 
+    'O': 1, 
+    'P': 3, 
+    'Q': 10, 
+    'R': 1, 
+    'S': 1, 
+    'T': 1, 
+    'U': 1, 
+    'V': 4, 
+    'W': 4, 
+    'X': 8, 
+    'Y': 4, 
+    'Z': 10
+}
+
 def draw_letters():
-    pass
+    # use random.sample to draw 10 letters from the 
+    # letter bank.
+
+    # get the lists of letter and count for each letter
+    # as list (preferred by random.sample)
+    letter_list = list(LETTER_POOL.keys())
+    letter_count = list(LETTER_POOL.values())
+
+    # sample draws k items from the list using counts 
+    # to determine the frequency of each item. 
+    # https://docs.python.org/3/library/random.html#random.sample
+    # requires python version 3.9 or greater 
+    this_hand = sample(letter_list, k=10, counts=letter_count)
+
+    # print('-----------------')
+    # print(this_hand)
+    # print('-----------------')
+
+
+    return this_hand
 
 def uses_available_letters(word, letter_bank):
-    pass
+    word = word.upper()
+
+    # create a counter for letter_bank
+    letter_bank_counts = Counter()
+
+    for letter in letter_bank:
+        letter_bank_counts[letter] += 1
+    # print("Loaded counter: ", letter_bank_counts)
+
+    for letter in list(word):
+        if letter_bank_counts[letter] == 0:
+            return False
+        else:
+            letter_bank_counts[letter] -= 1
+    
+    return True
+
+
+# print(uses_available_letters)
 
 def score_word(word):
-    pass
+    score_total = 0
+    extra_points = 8
+    word = word.upper()
+
+    letters_from_word = list(word)
+
+    if 7 <= len(word) <= 10:    
+        score_total += extra_points
+
+    for letter in letters_from_word: 
+        if letter in SCORE_CHART.keys():
+            score_total += SCORE_CHART[letter]
+
+    return(score_total)
+    
+
+def _high_score_key_helper(word):
+    # returns a tuple (score, has_10, -length)
+
+    cur_score = score_word(word)
+    word_len = -len(word)
+    has_10 = word_len == -10 # True if word length is 10.
+
+    result = (cur_score, has_10, word_len)
+    return result 
+
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+
+    # use our helper function to find the highest ranked
+    # word using max
+    high_word = max(word_list, key=_high_score_key_helper)
```

Comment:
Having sat with this a bit more, I'm going to go with I like this better than the explicit approach in the long run, and that my initial hesitation has more to do with the languages I've used before. But this is Python, so when in Rome...

My main hesitation come from the fact that in other languages, the way we tend to provide customization for functions like min, max, and sort comes from providing a comparison function. We explicitly write the logic to compare two instances of the data in the list. Python does support this style of customization through the use of an adapter type (functools.cmp_to_key) that can turn a comparison function into something that will work as the key function. But if we embrace the key function fully, it can effectively give us a compound key, comparing several aspects as needed.

It _still_ feels odd to me to compare True and False in a relative sense (Java, for instance does not allow this, but we could simply map true to 1 and false to 0), but Python definitely allows it.

So again, really clever implementation, and great use of the tools of the language!

Reply:
I got really confused trying to figure out the if/else tree and thought about putting it into a comparison function so that we could test how the logic works just word vs. word. But python3 doesn't have comparison functions anymore, just key functions, and somewhere tuple sorting clicked analogous to (LastName, FirstName, MiddleInitial). 

I agree that sorting on a Boolean feels a bit weird though.